### PR TITLE
Changes to excluded test failures when using mono runtime.

### PR DESCRIPTION
**Pull request by naricc**
These changes let us exclude known failures on the mono run time.  

They also fix some things that prevented the tests from being built in debug.

### CODE REVIEW

File: src/coreclr/tests/src/Common/test_dependencies/test_dependencies.csproj
Comment (trylek) [inline src/coreclr/tests/src/Common/test_dependencies/test_dependencies.csproj]: @jkoritzinsky - have you got any context why this needs changing? I don't see anything outright wrong here but it seems orthogonal to the task of enabling mono-specific test exclusions as such. @naricc, if this is a semantic change that is for whatever reason required for Mono builds, it should be made mono-conditional and ideally equipped with an explanatory comment why this logic differs between Mono and CoreCLR.

File: src/coreclr/tests/src/Directory.Build.targets
Comment (trylek) [inline src/coreclr/tests/src/Directory.Build.targets]: Same as above.

File: src/coreclr/build-test.sh
Comment (trylek) [inline src/coreclr/build-test.sh]: For better alignment with other conditional sections in issues.targets, could you rather expose a variable named something like RuntimeType with the value "coreclr" or "mono" and change the check in issues.targets to <code>'$(RuntimeType)' == 'mono'</code>?

**Review by trylek — CHANGES_REQUESTED** [review]
Looks mostly good, thanks for pulling this off! I would suggest a bit of mop-up work though:

1) I'd like to better understand why we're changing the library configuration as it seems orthogonal to this task as such;

2) Please consider following a similar pattern to we're using for other architecture, build flavor and OS-specific exclusions in issues.targets (see my comment regarding 'RuntimeType' variable);

3) What about the Windows script (build-test.cmd), are you planning to include it in a subsequent commit? The mono-specific exclusions wouldn't work without it on Windows even though probably nothing tragic would happen, the variable would just be undefined and expanded to an empty string and the Mono-specific exclusions would just be ignored.

Thanks

Tomas

File: src/coreclr/tests/src/Common/test_dependencies/test_dependencies.csproj
Comment (jkoritzinsky) [reply src/coreclr/tests/src/Common/test_dependencies/test_dependencies.csproj]: I don't know why we'd want to change this. The CoreCLR tests should always use the Release build of Libraries (there's no reason for them to use debug).

Comment (naricc) [reply src/coreclr/tests/src/Common/test_dependencies/test_dependencies.csproj]: @trylek 

I found that if its hard coded for Release here, the build-test will always check that some files exist in /artifacts/...release... directory, even if you have built everything in debug. With these two changes, I can build in debug and everything works.   

I can move them to a seperate PR if you like, since they are orthoganal. But I would like to fix it, unless there is some reason not to. I found it really confusing when trying to come up to speed on this, and it will probably trip some one else up in the future.

File: src/coreclr/tests/src/Directory.Build.targets
Comment (naricc) [reply src/coreclr/tests/src/Directory.Build.targets]: Ditto my reply above on this change.

File: src/coreclr/tests/src/Common/test_dependencies/test_dependencies.csproj
Comment (naricc) [reply src/coreclr/tests/src/Common/test_dependencies/test_dependencies.csproj]: Alternatively, we could just make build-tests.sh not accept the -Debug argument; that would also remove the confusion.

File: src/coreclr/build-test.sh
Comment (naricc) [reply src/coreclr/build-test.sh]: Ok, that makes sense. I will do it that way.

File: src/coreclr/tests/src/Common/test_dependencies/test_dependencies.csproj
Comment (jkoritzinsky) [reply src/coreclr/tests/src/Common/test_dependencies/test_dependencies.csproj]: We want a way to be able to build the tests in Debug but still always use the release libraries. I'd rather have this change in a separate PR so more of the CoreCLR teams can chime in on how they would like this to work.

cc: @jashook

Comment (naricc) [reply src/coreclr/tests/src/Common/test_dependencies/test_dependencies.csproj]: Ok; I will take it out of this PR. It sounds like maybe a more comprehensive change is needed later.

**Comment by naricc** [comment]
@trylek I planned to make the change to the .cmd file in a seperate PR.  I don't actually have a development environment that can run .cmd files and didn't want to hold it up more while I set that up. Since it does not break anything, and the .sh file is probably more important for mono, that seems reasonable to me.

**Review by trylek — APPROVED** [review]
LGTM, thank you!

**Comment by SamMonoRT** [comment]
How do we exclude tests based on a particular mono config like for example running tests with interpreter or running with llvm ?

**Comment by trylek** [comment]
I guess that the existing mechanism can be further refined. The good news is that the issues.targets file kicks in after the actual managed test build, in the final phase of XUnit wrapper creation, so if there's for instance a reasonable local scenario for running multiple such mono configs in a row, it doesn't require rebuilding the tests. I treat Nathan's work as a pilot change in this direction that many others can follow along the same lines.

File: src/coreclr/tests/issues.targets
Comment (jashook) [inline src/coreclr/tests/issues.targets]: What OS/Arch is this disabling tests for?

Comment (jashook) [reply src/coreclr/tests/issues.targets]: Are these tests that are completely incompatible with the mono runtime?

Comment (trylek) [reply src/coreclr/tests/issues.targets]: That's my understanding; maybe "completely incompatible" is a bit too strong statement, I guess that some of them may [and perhaps should] be fixed to be mono-friendly. Not sure about bits where the two runtimes fundamentally differ like GC.

**Review by jashook — CHANGES_REQUESTED** [review]
I personally would like to see the exclusions mapped into groups that are easy to understand which tests are disabled for a specifc runtime/arch/os.

As it is, it is hard to understand what os/arch the mono runtime tests are excluded under. My suggestion would be to add runtimename = coreclr and add this as check to each of the existing exclusion lists.

Basically, by adding mono, we are going to go from filtering on os/arch to runtime/os/arch and I would like to see that clearly annotated, as this file has been the source of weird problems in the past, where we suddenly stop building tests and we miss it for a while.

File: src/coreclr/tests/issues.targets
Comment (jashook) [reply src/coreclr/tests/issues.targets]: Correct, the question is really to understand what this when this new bucket is actually attempting to disable and why.

Comment (jashook) [reply src/coreclr/tests/issues.targets]: For example, if these are just the known set of tests that fail on mono/x64/linux then I would like us to annotate it as such.

**Comment by trylek** [comment]
We are actually also filtering for Crossgen2, Simon just merged that in ;-).

**Comment by trylek** [comment]
For issues.targets in general I think you're right there are two smoking guns we should tackle somehow:

1) Running some regular cleanup bot that would throw away tests blocked on issues that have been closed. This makes sure that a test doesn't remain disabled for another year just because the person who fixed the primary issue forgot, didn't realize or didn't even know he should remove an issues.targets exclusion.

2) Getting rid of pseudo-items like "Needs triage" (I admit this change adds quite a few but we also have pre-existing ones) and replacing them with real actionable items conforming to (1).

My personal feeling is that we shouldn't insist on doing that as part of this PR to unblock the primary task of "making Mono tests green". As this is a new addition, I would initially ping @marek-safar for their action plan on triaging / addressing these issues and follow up with @jaredpar on the longer-term issues.targets policy. In practice it amounts to slightly more than 200 (or 10%) of Pri0 tests and it doesn't regress CoreCLR testing in any manner.

File: src/coreclr/tests/issues.targets
Comment (trylek) [reply src/coreclr/tests/issues.targets]: I think that should be part of the triage the test claim to need, the important bit is to make sure this triage actually happens at some point. Perhaps the hypothetical issue cleanup bot could also emit warnings for tests without valid work item ID's so that they could be brought up at infra sync-ups and such.

**Comment by trylek** [comment]
Actually, @naricc - can you confirm whether the exclusions cover Pri0 or Pri1 tests? I mean, if these are just Pri0 tests, we'll likely need about 800 more exclusions for the Pri1 tests, otherwise outerloop runs will remain red.

**Comment by naricc** [comment]
@trylek 

This is just the priority 0 tests.

File: src/coreclr/tests/issues.targets
Comment (naricc) [reply src/coreclr/tests/issues.targets]: Many of these I believe will fail on Mono across all architectures, although I have chiefly been testing on OSX/x64. I am definitely aware that I am going to have to create some more categories, but it is a process.

I don't object to adding a RunTime == CoreCLR check to the existing exclusions, but I will have to figure out which of those also fail on Mono and which do not.

As for creating the actual issues, I plan to create them this week.

**Comment by marek-safar** [comment]
I see a few purposes of this "get something going" PR and it should go in with following follow up for @SamMonoRT to handle.

- getting this hooked up into PRs testing (this included figuring out how change in coreclr folder will force mono retest)
- create an issue for better categorization of issues in `issues.targets` (the current setup does not look very scalable to me).
- Create issue for each failing tests on Mono configuration as we did for set/libraries tests

**Review by marek-safar — APPROVED** [review]

File: src/coreclr/build-test.sh
Comment (akoeplinger) [inline src/coreclr/build-test.sh]: We should reuse the `RuntimeFlavor` property we already use in other parts of the build to decide between Mono/CoreCLR instead of inventing another property here. Can be done in a follow-up PR though :)

Comment (jaredpar) [reply src/coreclr/build-test.sh]: Agree. If we're doing it as a follow up PR though we should file an issue to track the work so it doesn't get lost.

**Comment by SamMonoRT** [comment]
I'll be creating 4 issues : 
1. Investigate and hook up running the with PRs/CI, making sure any change in CoreCLR or Mono triggers these test run on Mono
2. Refactor the issues.target file to better categorize for different OS/architecture specific failures
3. Remove RuntimeName property and use existing RuntimeFlavor instead.
4. Update the issues.target file as issues are created for each failing excluded test

**Comment by naricc** [comment]
@jashook So from the discussion it sounds like we decided to address the issues in follow up PRs. So can you unblock this unless you have additonal issues?

 I did update the comment in issues.target to make it more clear what we are attempting to exclude

**Review by jashook — APPROVED** [review]

**Pull request merged**
